### PR TITLE
8276673: Optimize abs operations in C2 compiler

### PR DESCRIPTION
**Pull request by fg1417**
The patch aims to help optimize Math.abs() mainly from these three parts:
1) Remove redundant instructions for abs with constant values
2) Remove redundant instructions for abs with char type
3) Convert some common abs operations to ideal forms

1. Remove redundant instructions for abs with constant values

If we can decide the value of the input node for function Math.abs()
at compile-time, we can substitute the Abs node with the absolute
value of the constant and don't have to calculate it at runtime.

For example,
  int[] a
  for (int i = 0; i < SIZE; i++) {
    a[i] = Math.abs(-38);
  }

Before the patch, the generated code for the testcase above is:
...
  mov   w10, #0xffffffda
  cmp   w10, wzr
  cneg  w17, w10, lt
  dup   v16.8h, w17
...
After the patch, the generated code for the testcase above is :
...
  movi  v16.4s, #0x26
...

2. Remove redundant instructions for abs with char type

In Java semantics, as the char type is always non-negative, we
could actually remove the absI node in the C2 middle end.

As for vectorization part, in current SLP, the vectorization of
Math.abs() with char type is intentionally disabled after
JDK-8261022 because it generates incorrect result before. After
removing the AbsI node in the middle end, Math.abs(char) can be
vectorized naturally.

For example,

  char[] a;
  char[] b;
  for (int i = 0; i < SIZE; i++) {
    b[i] = (char) Math.abs(a[i]);
  }

Before the patch, the generated assembly code for the testcase
above is:

B15:
  add   x13, x21, w20, sxtw #1
  ldrh  w11, [x13, #16]
  cmp   w11, wzr
  cneg  w10, w11, lt
  strh  w10, [x13, #16]
  ldrh  w10, [x13, #18]
  cmp   w10, wzr
  cneg  w10, w10, lt
  strh  w10, [x13, #18]
  ...
  add   w20, w20, #0x1
  cmp   w20, w17
  b.lt  B15

After the patch, the generated assembly code is:
B15:
  sbfiz x18, x19, #1, #32
  add   x0, x14, x18
  ldr   q16, [x0, #16]
  add   x18, x21, x18
  str   q16, [x18, #16]
  ldr   q16, [x0, #32]
  str   q16, [x18, #32]
  ...
  add   w19, w19, #0x40
  cmp   w19, w17
  b.lt  B15

3. Convert some common abs operations to ideal forms

The patch overrides some virtual support functions for AbsNode
so that optimization of gvn can work on it. Here are the optimizable
forms:

a) abs(0 - x) => abs(x)

Before the patch:
  ...
  ldr   w13, [x13, #16]
  neg   w13, w13
  cmp   w13, wzr
  cneg  w14, w13, lt
  ...
After the patch:
  ...
  ldr   w13, [x13, #16]
  cmp   w13, wzr
  cneg  w13, w13, lt
  ...

b) abs(abs(x))  => abs(x)

Before the patch:
  ...
  ldr   w12, [x12, #16]
  cmp   w12, wzr
  cneg  w12, w12, lt
  cmp   w12, wzr
  cneg  w12, w12, lt
  ...
After the patch:
  ...
  ldr   w13, [x13, #16]
  cmp   w13, wzr
  cneg  w13, w13, lt
  ...

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276673](https://bugs.openjdk.java.net/browse/JDK-8276673): Optimize abs operations in C2 compiler

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Jie Fu](https://openjdk.java.net/census#jiefu) (@DamonFool - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6755/head:pull/6755` \
`$ git checkout pull/6755`

Update a local copy of the PR: \
`$ git checkout pull/6755` \
`$ git pull https://git.openjdk.java.net/jdk pull/6755/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6755`

View PR using the GUI difftool: \
`$ git pr show -t 6755`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6755.diff">https://git.openjdk.java.net/jdk/pull/6755.diff</a>

</details>

### CODE REVIEW

**Comment by fg1417** [comment]
The PR optimizes abs operations in the C2 middle end. Can I have your review please?

**Comment by DamonFool** [comment]
> The PR optimizes abs operations in the C2 middle end. Can I have your review please?

So what's the performance data before and after this patch?
Does it also benefit on x86?

It would be better to provide a jmh micro benchmark.
Thanks.

**Comment by fg1417** [comment]
> > The PR optimizes abs operations in the C2 middle end. Can I have your review please?
> 
> So what's the performance data before and after this patch? Does it also benefit on x86?
> 
> It would be better to provide a jmh micro benchmark. Thanks.

Thanks, @DamonFool . Yes, it's supposed to benefit all archs. 
For example, here is the performance data on x86.

Before the patch:
Benchmark                 (seed)   Mode  Cnt       Score       Error   Units
MathBench.absConstantInt       0  thrpt    5  291960.380 ± 10724.572  ops/ms

After the patch:
Benchmark                 (seed)   Mode  Cnt       Score      Error   Units
MathBench.absConstantInt       0  thrpt    5  336271.533 ± 3778.210  ops/ms

**Comment by DamonFool** [comment]
Hi @fg1417 ,

Thanks for your update.

Now I see that you are trying to optimize the following three abs() patterns:
 1) Math.abs(-38)
 2) (char) Math.abs((char) c)
 3) Math.abs(0 - x)

But did you see these code patterns in real programs?
I'm a bit worried that we just improve the complexity of C2 with (almost) no performance gain in the real world.
Thanks.

File: src/hotspot/share/opto/subnode.cpp [hunk 1]
```diff
@@ -1839,6 +1839,72 @@ bool BoolNode::is_counted_loop_exit_test() {
...
+  const Type* t1 = phase->type(in(1));
+  if (t1 == Type::TOP) return Type::TOP;
+
+  switch (t1->base()) {
+  case Type::Int: {
+    const TypeInt* ti = t1->is_int();
+    if (ti->is_con()) {
+      // Special case for min_jint: Math.abs(min_jint) = min_jint.
+      // Do not use C++ abs() for min_jint to avoid undefined behavior.
+      return (ti->is_con(min_jint)) ? TypeInt::MIN : TypeInt::make(abs(ti->get_con()));
```
Comment (theRealAph) [inline src/hotspot/share/opto/subnode.cpp:1854]:
```suggestion
      return TypeInt::make(uabs(ti->get_con());
```
We have uabs() for julong and unsigned int.

**Comment by fg1417** [comment]
> Hi @fg1417 ,
> 
> Thanks for your update.
> 
> Now I see that you are trying to optimize the following three abs() patterns:
> 
> 1. Math.abs(-38)
> 2. (char) Math.abs((char) c)
> 3. Math.abs(0 - x)
> 
> But did you see these code patterns in real programs? I'm a bit worried that we just improve the complexity of C2 with (almost) no performance gain in the real world. Thanks.

Thanks for your review, @DamonFool . I really understand your concern.

In terms of complexity, the change only involves AbsNode and doesn’t modify any other code part.  I don’t think it will make C2 more complex. 

As for performance gain in the real world, as we all know, the ability of GVN to optimize a node often depends on the optimized result of its input nodes. For example, if the input node of one AbsNode is recognized as a constant after last round of GVN optimization, now, we can optimize abs(constant) to a simple constant value. Like C2 did in https://github.com/openjdk/jdk/blob/0bddd8af61b6c731f16b857c09de57ceefd72d06/src/hotspot/share/opto/subnode.cpp#L55, we may not see -(-x) or (x+y)-y in any java program directly, but it’s possible after C2 optimization.  Whether the optimization to sub or to abs is trivial, low-cost but useful. Why not apply it :)

Math.abs(-38) ,  (char) Math.abs((char) c) and Math.abs(0 - x) are just conformance testcases. As you said, maybe nobody writes these cases in the real world. These testcases are just simulating all possible scenarios that AbsNode may meet, to guarantee the correctness of the optimization.

What do you think :)

Thanks.

**Comment by DamonFool** [comment]
> > Hi @fg1417 ,
> > Thanks for your update.
> > Now I see that you are trying to optimize the following three abs() patterns:
> > 
> > 1. Math.abs(-38)
> > 2. (char) Math.abs((char) c)
> > 3. Math.abs(0 - x)
> > 
> > But did you see these code patterns in real programs? I'm a bit worried that we just improve the complexity of C2 with (almost) no performance gain in the real world. Thanks.
> 
> Thanks for your review, @DamonFool . I really understand your concern.
> 
> In terms of complexity, the change only involves AbsNode and doesn’t modify any other code part. I don’t think it will make C2 more complex.
> 
> As for performance gain in the real world, as we all know, the ability of GVN to optimize a node often depends on the optimized result of its input nodes. For example, if the input node of one AbsNode is recognized as a constant after last round of GVN optimization, now, we can optimize abs(constant) to a simple constant value. Like C2 did in
> 
> https://github.com/openjdk/jdk/blob/0bddd8af61b6c731f16b857c09de57ceefd72d06/src/hotspot/share/opto/subnode.cpp#L55
> 
> , we may not see -(-x) or (x+y)-y in any java program directly, but it’s possible after C2 optimization. Whether the optimization to sub or to abs is trivial, low-cost but useful. Why not apply it :)
> Math.abs(-38) , (char) Math.abs((char) c) and Math.abs(0 - x) are just conformance testcases. As you said, maybe nobody writes these cases in the real world. These testcases are just simulating all possible scenarios that AbsNode may meet, to guarantee the correctness of the optimization.
> 
> What do you think :)
> 
> Thanks.

Then, shall we also opt cases like `Math.abs(-1 * x)`, `Math.abs(x / (-1))`, and so on?
Thanks.

**Comment by fg1417** [comment]
> > > Hi @fg1417 ,
> > > Thanks for your update.
> > > Now I see that you are trying to optimize the following three abs() patterns:
> > > 
> > > 1. Math.abs(-38)
> > > 2. (char) Math.abs((char) c)
> > > 3. Math.abs(0 - x)
> > > 
> > > But did you see these code patterns in real programs? I'm a bit worried that we just improve the complexity of C2 with (almost) no performance gain in the real world. Thanks.
> > 
> > 
> > Thanks for your review, @DamonFool . I really understand your concern.
> > In terms of complexity, the change only involves AbsNode and doesn’t modify any other code part. I don’t think it will make C2 more complex.
> > As for performance gain in the real world, as we all know, the ability of GVN to optimize a node often depends on the optimized result of its input nodes. For example, if the input node of one AbsNode is recognized as a constant after last round of GVN optimization, now, we can optimize abs(constant) to a simple constant value. Like C2 did in
> > https://github.com/openjdk/jdk/blob/0bddd8af61b6c731f16b857c09de57ceefd72d06/src/hotspot/share/opto/subnode.cpp#L55
> > 
> > , we may not see -(-x) or (x+y)-y in any java program directly, but it’s possible after C2 optimization. Whether the optimization to sub or to abs is trivial, low-cost but useful. Why not apply it :)
> > Math.abs(-38) , (char) Math.abs((char) c) and Math.abs(0 - x) are just conformance testcases. As you said, maybe nobody writes these cases in the real world. These testcases are just simulating all possible scenarios that AbsNode may meet, to guarantee the correctness of the optimization.
> > What do you think :)
> > Thanks.
> 
> Then, shall we also opt cases like `Math.abs(-1 * x)`, `Math.abs(x / (-1))`, and so on? Thanks.

Hi, @DamonFool .

Actually, the cases you listed above, `Math.abs(-1 * x)`, `Math.abs(x / (-1))`, are covered by the optimized pattern, `Math.abs(0-x)`.

In C2, `-1 * x` is going to be `0 – x` after GVN optimization in MulNode. C2 takes `-1 * x` as `0 – (1*x)` firstly, and Identity here, https://github.com/openjdk/jdk/blob/0bddd8af61b6c731f16b857c09de57ceefd72d06/src/hotspot/share/opto/mulnode.cpp#L50, will combine `1*x` to x. After that, `0 – x` matched the pattern that AbsNode can recognize. `Math.abs(x / (-1))` , too. As I mentioned before, the AbsNode optimization doesn’t work as a standalone pass and it’s combined very closely to its input nodes. Instruction sequence changes as GVN repeats and our ideal pattern may occur.

Your cases also help proved that these several patterns, like abs(0-x) and abs(positive_value), are very fundamental and common. That’s why we choose them.

Thanks.

**Comment by DamonFool** [comment]
> Your cases also help prove that these several patterns, like abs(0-x) and abs(positive_value), are very fundamental and common.

I don't think so.
In real programs, I will never write code like `Math.abs(0 - x)`, `Math.abs(-1 * x)` and `Math.abs(x / (-1))`.

How about improving your micro-bechmark to show the performance gain?
To be honest, benchmarking with `Math.abs(-3)` seems strange to me since I don't think people will write that code.
So I would suggest writing a jmh test which may be used in real programs.
Thanks.

File: src/hotspot/share/opto/subnode.cpp (hunk 1 above)
Comment (fg1417) [reply src/hotspot/share/opto/subnode.cpp]: Thanks for your review. Fixed.

**Comment by DamonFool** [comment]
I discussed this opt with @theRealAph offline.

To clarify from my point of view:
 1. I have no objection to this PR.
 2. I'd like to see a benchmark which people would write in real programs.
 3. But if the OpenJDK experts think it's already good enough, please go ahead.

Thanks.

**Comment by fg1417** [comment]
> I discussed this opt with @theRealAph offline.
> 
> To clarify from my point of view:
> 
> 1. I have no objection to this PR.
> 2. I'd like to see a benchmark which people would write in real programs.
> 3. But if the OpenJDK experts think it's already good enough, please go ahead.
> 
> Thanks.

Hi, @DamonFool 

I ran the jtreg test internally with some logging info to verify if the optimization works in real java program. The results shows that these patterns are hit in the following cases:

•	java/lang/StackWalker/LocalsAndOperands.java#id0
•	java/lang/StackWalker/LocalsAndOperands.java#id1
•	java/lang/invoke/LFCaching/LFSingleThreadCachingTest.java
•	java/util/concurrent/tck/JSR166TestCase.java
•	javax/management/timer/MissingNotificationTest.java
•	jdk/incubator/vector/Double128VectorTests.java
•	jdk/incubator/vector/Double256VectorTests.java
•	jdk/incubator/vector/Double512VectorTests.java
•	jdk/incubator/vector/Double64VectorTests.java
•	jdk/incubator/vector/DoubleMaxVectorTests.java
•	jdk/incubator/vector/Float128VectorTests.java
•	jdk/incubator/vector/Float256VectorTests.java
•	jdk/incubator/vector/Float512VectorTests.java
•	jdk/incubator/vector/Float64VectorTests.java
•	jdk/incubator/vector/FloatMaxVectorTests.java
•	jdk/incubator/vector/Vector128ConversionTests.java
•	jdk/incubator/vector/Vector256ConversionTests.java
•	jdk/incubator/vector/Vector64ConversionTests.java#id0
•	jdk/incubator/vector/VectorMaxConversionTests.java

It’s not easy to identify these patterns from original java code by our eyes. Since the added code lines are hit, the patterns must occur after many rounds of optimization. Definitely, it benefits all platform, whether x86 or aarch64.

 As for the current benchmark, it’s not to show the real performance gain but to illustrate that the opto benefits x86 as well in case you wonder. If you need a real java program, the case won’t be light-weight or straightforward. Maybe I can’t provide you with a satisfying micro benchmark.

Thanks.

**Comment by DamonFool** [comment]
> > I discussed this opt with @theRealAph offline.
> > To clarify from my point of view:
> > 
> > 1. I have no objection to this PR.
> > 2. I'd like to see a benchmark which people would write in real programs.
> > 3. But if the OpenJDK experts think it's already good enough, please go ahead.
> > 
> > Thanks.
> 
> Hi, @DamonFool
> 
> I ran the jtreg test internally with some logging info to verify if the optimization works in real java program. The results shows that these patterns are hit in the following cases:
> 
> • java/lang/StackWalker/LocalsAndOperands.java#id0 • java/lang/StackWalker/LocalsAndOperands.java#id1 • java/lang/invoke/LFCaching/LFSingleThreadCachingTest.java • java/util/concurrent/tck/JSR166TestCase.java • javax/management/timer/MissingNotificationTest.java • jdk/incubator/vector/Double128VectorTests.java • jdk/incubator/vector/Double256VectorTests.java • jdk/incubator/vector/Double512VectorTests.java • jdk/incubator/vector/Double64VectorTests.java • jdk/incubator/vector/DoubleMaxVectorTests.java • jdk/incubator/vector/Float128VectorTests.java • jdk/incubator/vector/Float256VectorTests.java • jdk/incubator/vector/Float512VectorTests.java • jdk/incubator/vector/Float64VectorTests.java • jdk/incubator/vector/FloatMaxVectorTests.java • jdk/incubator/vector/Vector128ConversionTests.java • jdk/incubator/vector/Vector256ConversionTests.java • jdk/incubator/vector/Vector64ConversionTests.java#id0 • jdk/incubator/vector/VectorMaxConversionTests.java
> 
> It’s not easy to identify these patterns from original java code by our eyes. Since the added code lines are hit, the patterns must occur after many rounds of optimization. Definitely, it benefits all platforms, whether x86 or aarch64.
> 
> As for the current benchmark, it’s not to show the real performance gain but to illustrate that the opto benefits x86 as well in case you wonder. If you need a real java program, the case won’t be light-weight or straightforward. Maybe I can’t provide you with a satisfying micro benchmark.
> 
> Thanks.

Good news!

But can you show us an example with more detailed analysis which pattern is applied in the test? 
Thanks.

File: test/hotspot/jtreg/compiler/c2/TestAbs.java [hunk 2]
```diff
@@ -21,17 +21,151 @@
...
  *
  * @run main/othervm -XX:-TieredCompilation -Xbatch -XX:CompileOnly=java.lang.Math::abs compiler.c2.TestAbs
+ * @run main/othervm -XX:-TieredCompilation compiler.c2.TestAbs
  */
+
+package compiler.c2;
+import jdk.test.lib.Asserts;
+
 public class TestAbs {
+    private static int SIZE = 500;
```
Comment (DamonFool) [inline test/hotspot/jtreg/compiler/c2/TestAbs.java:39]: Not used?

File: test/hotspot/jtreg/compiler/c2/TestAbs.java [hunk 3]
```diff
@@ -21,17 +21,151 @@
...
+        Asserts.assertEquals(Integer.MIN_VALUE, Math.abs(Integer.MIN_VALUE));
+        Asserts.assertEquals(Integer.MAX_VALUE, Math.abs(-Integer.MAX_VALUE));
+
+        // Test abs(constant) optimization for long
+        Asserts.assertEquals(Long.MAX_VALUE, Math.abs(Long.MAX_VALUE));
+        Asserts.assertEquals(Long.MIN_VALUE, Math.abs(Long.MIN_VALUE));
+        Asserts.assertEquals(Long.MAX_VALUE, Math.abs(-Long.MAX_VALUE));
+
+        // Test abs(constant) optimization for float
+        Asserts.assertEquals(Float.NaN, Math.abs(Float.NaN));
```
Comment (DamonFool) [inline test/hotspot/jtreg/compiler/c2/TestAbs.java:114]: I would suggest something like:
```
assertTrue(Float.isNaN(Math.abs(Float.NaN)))
```

File: test/hotspot/jtreg/compiler/c2/TestAbs.java [hunk 4]
```diff
@@ -21,17 +21,151 @@
...
+        Asserts.assertEquals(Double.POSITIVE_INFINITY, Math.abs(Double.POSITIVE_INFINITY));
+        Asserts.assertEquals(0.0, Math.abs(0.0));
+        Asserts.assertEquals(0.0, Math.abs(-0.0));
+        Asserts.assertEquals(Double.MAX_VALUE, Math.abs(Double.MAX_VALUE));
+        Asserts.assertEquals(Double.MIN_VALUE, Math.abs(Double.MIN_VALUE));
+        Asserts.assertEquals(Double.MAX_VALUE, Math.abs(-Double.MAX_VALUE));
+        Asserts.assertEquals(Double.MIN_VALUE, Math.abs(-Double.MIN_VALUE));
+    }
+
+    private static void testAbsTransformInt(int[] a) {
```
Comment (DamonFool) [inline test/hotspot/jtreg/compiler/c2/TestAbs.java:136]: If you want to verify C2's transformation, probably we should use C2's IR test framework.

**Comment by fg1417** [comment]
> But can you show us an example with more detailed analysis which pattern is applied in the test? Thanks.

Hi, @DamonFool 

For example, `jdk/incubator/vector/Float512VectorTests.java` calls `java.lang.FdLibm$Hypot::compute`. You can check math classes in Fdlibm.java, like `Hypot` or a common one `pow`, which call `Math.abs()`. After inline and many rounds of optimizations, such as constant propagation, the input value of `Math.abs()` in the `Pow::compute` is probably constant or `(0-x)`. We can optimize it using this patch.

I learnt the optimization technique and the way to verify from the patch of my colleague, 
[https://github.com/openjdk/jdk/pull/2776](url)
The similar question was answered by Tobias in the conversation, and you can refer to it.

Thanks.

**Comment by fg1417** [comment]
> But can you show us an example with more detailed analysis which pattern is applied in the test? Thanks.

Hi, @DamonFool 

For example, `jdk/incubator/vector/Float512VectorTests.java` calls `java.lang.FdLibm$Hypot::compute`. You can check math classes in `Fdlibm.java`, like `Hypot` or a common one `Pow`, which calls `Math.abs()`. After inline and many optimizations, such as constant propagation., the input value of `Math.abs()` is probably constant or `(0-x)`. We can optimize it with this patch.

I learnt the optimization technique from the patch of my colleague,
[ https://github.com/openjdk/jdk/pull/2776](url)
The similar question was answered by Tobias in the conversation, and you can refer to it.

Thanks.

**Comment by fg1417** [comment]
> But can you show us an example with more detailed analysis which pattern is applied in the test? Thanks.
Hi, @DamonFool 

For example, `jdk/incubator/vector/Float512VectorTests.java` calls `java.lang.FdLibm$Hypot::compute`. You can check math classes in `Fdlibm.jav`a, like `Hypot`  or a common one `Pow`, which calls `Math.abs()`. After inline and many optimizations, such as constant propagation., the input value of `Math.abs()` is probably constant or `(0-x)`. We can optimize it using this patch.

I learnt the optimization technique from the patch of my colleague, https://github.com/openjdk/jdk/pull/2776#issuecomment-789756226
The similar question was answered by Tobias in the conversation, and you can refer to it.

Thanks.

**Comment by DamonFool** [comment]
> > But can you show us an example with more detailed analysis which pattern is applied in the test? Thanks.
> 
> Hi, @DamonFool
> 
> For example, `jdk/incubator/vector/Float512VectorTests.java` calls `java.lang.FdLibm$Hypot::compute`. You can check math classes in `Fdlibm.java`, like `Hypot` or a common one `Pow`, which calls `Math.abs()`. After inline and many optimizations, such as constant propagation, the input value of `Math.abs()` is probably constant or `(0-x)`. We can optimize it using this patch.
> 
> I learnt the optimization technique from the patch of my colleague, [#2776 (comment)](https://github.com/openjdk/jdk/pull/2776#issuecomment-789756226) The similar question was answered by Tobias in the conversation, and you can refer to it.
> 
> Thanks.

Very good!
You had proved that these patterns do exist in C2's opt passes, so this patch makes sense to me.
Thanks.

File: src/hotspot/share/opto/subnode.cpp [hunk 5]
```diff
@@ -1839,6 +1839,68 @@ bool BoolNode::is_counted_loop_exit_test() {
...
+
+//------------------------------Ideal------------------------------------------
+Node* AbsNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  Node* in1 = in(1);
+  // Convert "abs(0-x)" into "abs(x)"
+  if (in1->is_Sub() && phase->type(in1->in(1))->is_zero_type()) {
+    set_req(1, in1->in(2));
+    PhaseIterGVN* igvn = phase->is_IterGVN();
+    if (igvn) {
+      igvn->_worklist.push(in1);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/subnode.cpp:1897]: Why is that needed? Because `in1` could become dead? You should use `set_req_X` above.

**Review by TobiHartmann — COMMENTED** [review]
Looks good to me otherwise (expect the points about the test that @DamonFool already raised).

File: test/hotspot/jtreg/compiler/c2/TestAbs.java (hunk 4 above)
Comment (fg1417) [reply test/hotspot/jtreg/compiler/c2/TestAbs.java]: Done. Thanks.

File: test/hotspot/jtreg/compiler/c2/TestAbs.java (hunk 3 above)
Comment (fg1417) [reply test/hotspot/jtreg/compiler/c2/TestAbs.java]: Done. Thanks.

File: src/hotspot/share/opto/subnode.cpp (hunk 5 above)
Comment (fg1417) [reply src/hotspot/share/opto/subnode.cpp]: Done. Thanks.

File: test/hotspot/jtreg/compiler/c2/TestAbs.java (hunk 2 above)
Comment (fg1417) [reply test/hotspot/jtreg/compiler/c2/TestAbs.java]: Done. Thanks.

**Comment by fg1417** [comment]
Thanks for your review, @DamonFool @TobiHartmann . I fixed all your points mentioned above.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by DamonFool — APPROVED** [review]
LGTM
Thanks for the update.
